app/components/dialog.tsx

```diff
@@ -0,0 +1,41 @@
+import {useCallback, useRef} from 'react'
+
+type Props = {
+  children: any
+  onClose: (e: any) => void
+  [k: string]: any
+}
+
+export default function Dialog({children, onClose}: Props) {
+  // React magic to get the actual DOM element once it's been mounted
+  const dialogRef = useRef<HTMLDialogElement>(null)
+  const dialogSetter = useCallback((node: HTMLDialogElement) => {
+    if (node !== null) {
+      // Show the dialog this way, in order for it to automatically handle escapes
+      node.showModal()
+      dialogRef.current = node
+    }
+  }, [])
+
+  // Calculate whether the given click event was outside of the dialog, and if so close it
+  const closeIfOutside = (e: React.MouseEvent<HTMLElement>) => {
+    // If there isn't any dialog displayed, then the click obviously wasn't in it
+    if (dialogRef?.current == null) return
+
+    const rect = dialogRef.current.getBoundingClientRect()
+    const isInDialog =
+      rect.top <= e.clientY &&
+      e.clientY <= rect.top + rect.height &&
+      rect.left <= e.clientX &&
+      e.clientX <= rect.left + rect.width
+    if (!isInDialog) {
+      dialogRef.current.close()
+    }
+  }
+
+  return (
+    <dialog ref={dialogSetter} className="dialog" onClose={onClose} onClick={closeIfOutside}>
```

Comment:
I wanted to use the HTML5 dialog element. It can simply have an 'open' attribute to avoid all this ref magic, but the spec recommends calling `showModal()`, and it also handles escapes etc nicely if done this way